PR for issues #131, #2438, #366

Zegveld opened the pull request:
#131 #2438 #366 : Added support for SubClassMapping annotations. This will allow for hierarchy mappings.

if a parent mapping method is annotated with @SubClassMapping it will now generate an instanceof check inside the parent mapping and generate the subclass mappings if they are not manually defined. See issues #131 #2438 and #366 for more information.

Zegveld commented on processor/src/main/java/org/mapstruct/ap/internal/model/source/SubClassMappingOptions.java:
```diff
@@ -0,0 +1,135 @@
+
+        TypeMirror sourceSubClass = gem.sourceClass().getValue();
+        TypeMirror targetSubClass = gem.targetClass().getValue();
+        TypeMirror sourceParentType = method.getParameters().get( 0 ).asType();
```
What is the correct way to determine the sourceParentType? Currently assuming that the first argument of the method is the source that needs to be mapped to the target.

filiphr replied:
See https://github.com/mapstruct/mapstruct/pull/2512/#discussion_r697980035

Zegveld replied:
Replaced this with checking if any parameter matches. That way it doesn't matter which parameter the source parameter is.

filiphr replied:
I've left a comment about this in my last review. The current solution will also work for

```
@SubClassMapping(source = SubArg1, target = Target)
@SubClassMapping(source = SubArg2, target = Target)
Target map(SuperArg1 arg1, @Context SuperArg2 arg2);
```

However, it shouldn't. `@Context` arguments should not be treated as potential source.

Zegveld commented on processor/src/main/java/org/mapstruct/ap/internal/model/source/SubClassMappingOptions.java:
```diff
@@ -0,0 +1,135 @@
+    private static boolean isConsistent(SubClassMappingGem gem, ExecutableElement method, FormattingMessager messager,
+                                        TypeUtils typeUtils) {
+
+        if ( method.getParameters().isEmpty() || method.getReturnType().getKind() == TypeKind.VOID ) {
```
What is the correct way to determine the targetParentType? Currently assuming that the return value of the method is the target that needs to be mapped to.

filiphr replied:
See https://github.com/mapstruct/mapstruct/pull/2512/#discussion_r697980035

Zegveld replied:
This bit is still in it although I could probably retrieve it some other way, it does not really make sense to use SubClassMappings with the `@MappingTarget` annotation. Seeing as then both the source and the target need to be instanceof checked in order to call the correct method. Why not let the caller handle that, seeing as that one most likely will know what it is because it created the target object.
So leaving this as it is.

filiphr replied:
I've left a comment about how you can retrieve this in my last review.

filiphr commented:
Wow, thanks a lot for this PR @Zegveld. I haven't gotten around in reviewing this (picking the low hanging fruits first). However, we would look into this to make the 1.5.0.Beta2 release

filiphr commented on processor/src/main/resources/org/mapstruct/ap/internal/model/BeanMappingMethod.ftl:
```diff
@@ -24,6 +24,19 @@
     </#if>
 
+    <#if hasSubClassMappings()>
+        <#assign first = true />
+        <#list constructSubClassMappings(parameters) as subClass>
+            <#if !first>else</#if> if (${subClass.sourceArgument} instanceof <@includeModel object=subClass.sourceType/>) {
+                <@includeModel object=subClass.assignment existingInstanceMapping=existingInstanceMapping/>
+            }
+            <#assign first = false />
+        </#list>
```
no need to use the assignment for first. Check the [Accessing iteration state](https://freemarker.apache.org/docs/ref_directive_list.html#ref_list_accessing_state) for Freemarker. You can use `<#if subClass_index == 0>` instead. We are using that for the `afterMappingReferences`. I see that Freemarker has also added some more [Loop variable built-ins](https://freemarker.apache.org/docs/ref_builtins_loop_var.html), but we are not on that version yet 😄

filiphr commented on core/src/main/java/org/mapstruct/SubClassMapping.java:
```diff
@@ -0,0 +1,84 @@
+@Repeatable( value = SubClassMappings.class )
+@Retention( RetentionPolicy.CLASS )
+@Target( { ElementType.METHOD, ElementType.ANNOTATION_TYPE } )
+public @interface SubClassMapping {
+    /**
+     * @return the source subclass to check for before using the default mapping as fallback.
+     */
+    Class<?> sourceClass();
+
+    /**
+     * @return the target subclass to map the sourceSubClass to.
+     */
+    Class<?> targetClass();
+}
```
From an API point of view I think it would make sense to only have `source` and `target` instead of `sourceClass` and `targetClass`. The annotation already talks about `SubClass`, so it should be clear from that that `source` and `target` are classes.

WDYT @sjaakd?

filiphr commented on core/src/main/java/org/mapstruct/SubClassMapping.java:
```diff
@@ -0,0 +1,84 @@
+ * <p>
+ * The subclass to be mapped is to be specified via {@link #sourceClass()}.
+ * The subclass to map to is to be specified via {@link #targetClass()}.
+ * It currently is also required that a mapping method exist that has the {@link #targetClass()}
+ * as the return type and the {@link #sourceClass()} as the only argument.
```
Let's try to get rid of this requirement. I've left bunch of comments around that I think will help for this.

Zegveld replied:
The existing mapping is no longer required, but I'm still not certain about the mappings that use the update route. That might need some more work, but let's see how far we'll get.

filiphr replied:
I was more referring to the fact about "required that a mapping methods exists". That is not needed, since we can create an automatic method for it. We can mention that it does not support update methods though, no need to look into that for this PR

Zegveld replied:
I was referring to the same thing with the remark 'The existing mapping is no longer required'. :)
About the no support for update methods bit, I will add it (if not already there) to the documentation contained in the PR.

Note to self, this post requires a thumbs up when I've updated the documentation.

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:
```diff
@@ -560,6 +629,20 @@ else if ( !returnType.hasAccessibleConstructor() ) {
             return error;
         }
 
+        private boolean isReturnTypeAbstractOrcanBeConstructed(Type returnType) {
```
Small typo, "Orcan" to "OrCan"

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:
```diff
@@ -349,10 +368,60 @@ else if ( !method.isUpdateMethod() ) {
+                                     null,
+                                     null,
+                                     false );
+            Assignment assignment = ctx
+                                   .getMappingResolver()
+                                   .getTargetAssignment(
+                                       method,
+                                       null,
+                                       targetType,
+                                       FormattingParameters.EMPTY,
+                                       criteria,
+                                       rightHandSide,
+                                       null,
+                                           () -> forgeMapping( rightHandSide, sourceType, targetType ) );
```
I actually think that this invalidates the statement that a method is needed for the `SubClassMapping`.

Can we also have some tests where the configuration from the base mapping is passed to the forged mappings as well?

Zegveld replied:
This was a good idea, found a bug in the code with it.
Apparently the ForgedMethod that I created didn't get the MappingReferences of the parent method.
Fixing that as well now.

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/ForgedMethod.java:
```diff
@@ -347,7 +348,7 @@ public boolean isObjectFactory() {
     @Override
     public MappingMethodOptions getOptions() {
-        return basedOn.getOptions();
+        return MappingMethodOptions.getForgedMethodInheritedOptions( basedOn.getOptions() );
```
Why is this needed? Is this actually the fix for #2438? I would prefer if we can have that separate to reduce the surface area of a single PR.

Zegveld replied:
I think this change was to prevent infinite loops, so that the `@SubclassMapping` is not inherited. Will double-check this to be certain.

filiphr replied:
Makes sense, you need to remove the inherited `SubClassMapping`(s).  Let's create a field for this instead of creating it on every `getOptions` call. Those getOptions are called in a lot of places, so it might cause some performance problem.

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/processor/MapperCreationProcessor.java:
```diff
@@ -294,10 +296,20 @@ else if ( constructor.getParameters().size() == 1 ) {
         SortedSet<Type> extraImports = new TreeSet<>();
 
 
+        for ( SourceMethod method : methods ) {
+            for ( SubClassMappingOptions subClassMapping : method.getOptions().getSubClassMappings() ) {
+                Type type = typeFactory.getType( subClassMapping.getTargetClass() );
+                extraImports.add( type );
+                type = typeFactory.getType( subClassMapping.getSourceClass() );
+                extraImports.add( type );
+            }
+        }
```
I do not think that this is needed. The `SubClassMapping` already implements the `getImportTypes`, so you only need to call that for each `subClassMapping` in the `BeanMappingMethod#getImportTypes`

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:
```diff
@@ -1670,6 +1755,20 @@ public Type getReturnTypeToConstruct() {
+        return !subClasses.isEmpty();
+    }
+
+    public List<SubClassMapping> constructSubClassMappings(List<Parameter> parameters) {
+        subClasses.forEach( subClass -> subClass.updateWithParameters( parameters ) );
+        return subClasses;
+    }
```
We strive for the models to be immutable if possible. This code is invoked from the `BeanMappingMethod.ftl`. Can we do this in a different place and only get the `subClassMappings` here?

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingMethodOptions.java:
```diff
@@ -188,6 +200,8 @@ public void applyInheritedOptions(SourceMethod templateMethod, boolean isInverse
                 }
             }
 
+            // Do NOT inherit subclass mapping options!!!
```
Nice 😄

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/util/Message.java:
```diff
@@ -114,6 +114,9 @@
     ENUMMAPPING_ILLEGAL_TRANSFORMATION( "Illegal transformation for '%s' EnumTransformationStrategy. Error: '%s'." ),
 
+    SUBCLASSMAPPING_ILLEGAL_SUBCLASS("Class '%s' is not a subclass of '%s'."),
+    SUBCLASSMAPPING_METHOD_SIGNATURE_NOT_SUPPORTED("SubClassMapping annotation can not be used on Suppliers and Runnables."),
```
I do not think that this error message is clear. Having `void` return type means that the user is using update mappings, most likely we do not support that for now, so I would change the message to something like that.

By the way, you can do update mappings with a return type e.g. `Target update(@MappingTarget Target target, Source source)`

Zegveld replied:
So something like: `@SubClassMapping do not support update mappings.`?
Open for suggestions to make this message better, because the one above doesn't feel quite right either.

filiphr reviewed (commented):
Just finished with the review on this one. Really nice work @Zegveld. 

I've left some comments inline and I have some more general comments here.

I see that you are using the `_bugs` folder. We use this usually for actual bugs, not new features. I would suggest instead to use `subclassmapping` as top package and put your test there. You can then use different sub packages to show different use cases. In addition to that I am looking at the tests for #131 and #366. Correct me if I am wrong, but the only difference is the fact that in #131 `Vehicle` is not abstract, right?

Let's try to merge these things together and use the same source types and different target types. Perhaps you can also rename the source to `VehicleEntity` or somehow else, in order to avoid using full qualified names in the tests to make it more readable.

Can we also add some fixture tests, to see how the generated code looks like? As you know we strive to generate code which looks human readable 😄 .

Can we also add some erroneous tests to test the non happy paths for the new messages.

I am really excited in getting this in the upcoming 1.5. 

------
Something that could be good, but maybe we need to do it in a separate PR is to validate that the SubClassMapping make sense. For example if the first sub class mapping does not hide the second one.

e.g.

```
interface Vehicle { }

interface Car extends Vehicle { }

class SUV implements Car { }

class Sedan implements Car { }
```

And then if you defined sub class mapping like:

```
@SubClassMapping(source = Vehicle.class, target = ...)
@SubClassMapping(source = Car.class, target = ...)
```

Those subclass mappings will generate code which will never be executed.

e.g.

```
if (source instanceof Vehicle) {
    // ...
} else if (source instanceof Car) {
    // ...
}
```

After having written all of this, I think it makes sense to tackle it in a different PR. Let's do it like that, once we have this merged.

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/source/SubClassMappingOptions.java:
```diff
@@ -0,0 +1,135 @@
+                                                       FormattingMessager messager, TypeUtils typeUtils) {
+
+        SubClassMappingsGem subClassMappings = SubClassMappingsGem.instanceOn( method );
+        List<SubClassMappingGem> subClassMappingAnnotations;
+        if ( nonNull( subClassMappings ) ) {
+            subClassMappingAnnotations = subClassMappings.value().get();
+        }
+        else if ( nonNull( SubClassMappingGem.instanceOn( method ) ) ) {
+            subClassMappingAnnotations = Arrays.asList( SubClassMappingGem.instanceOn( method ) );
+        }
+        else {
+            return Collections.emptyList();
+        }
+
```
My comment here got lost. Let's do this in a similar way like we are doing it for `@Mapping` and `@Mappings`. We don't need to support meta annotations for now, but the location where we pick them up makes more sense in `MethodRetrieval#getMethodRequiringImplementation`. There you also have access to the source parameters, return type etc. This also answers your other 2 questions in this class I beleive.

Zegveld replied:
Made the flow for the `@SubClassMapping` and the `@SubClassMappings` equal to `@Mapping` and `@Mappings` with some refactoring. Added a common base class and two specific classes for the flow. This should also make it easier to add more implementations using the same route in the future. Something for the second review round.

Also added a sorter to solve inheritence ordering issues, the more specific a class the earlier it should be in the list. But not yet implemented it, might be that it will only get a real implementation when someone runs into this as an issue.

filiphr replied:
Yup I saw your refactoring. I like it a lot 😄 .

I've left a comment about the sorter in my last review. I don't think we need it for now.

filiphr commented:
By the way @Zegveld in https://github.com/mapstruct/mapstruct/issues/131#issuecomment-876330455 you are saying that you are not happy with the implementation route you took? Can you explain why? I personally thought that it is quite neat 😄 .

Zegveld commented:
> By the way @Zegveld in [#131 (comment)](https://github.com/mapstruct/mapstruct/issues/131#issuecomment-876330455) you are saying that you are not happy with the implementation route you took? Can you explain why? I personally thought that it is quite neat  .

Well that was before I made a lot more changes where in I made it quite neat ;)
Fixed it in this commit: https://github.com/rigd-loxia/mapstruct/commit/98c3d8e832d1277d1fa984a41c27ef168f8c1b17

Zegveld commented:
Only done the bits so far where I've added a thumbs up.
Still got a few review items to go, but that will have to wait till when I have time again later this week.

mjustin commented on core/src/main/java/org/mapstruct/SubClassMapping.java:
```diff
@@ -0,0 +1,82 @@
+@Repeatable( value = SubClassMappings.class )
+@Retention( RetentionPolicy.CLASS )
+@Target( { ElementType.METHOD, ElementType.ANNOTATION_TYPE } )
+public @interface SubClassMapping {
```
Since the term is the single word "[subclass](https://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-8.1.4)" in Java, wouldn't `SubclassMapping` be a more appropriate camel casing than `SubClassMapping`?

Zegveld replied:
I'm fine with it either way, I did not look into how compound words are camelcased in this project.

Here is the reason why I used `SubClassMapping` instead of `SubclassMapping`:
For compound words I look for example if I can put the text ` of a ` in between and retain the meaning.
So for `subclass` I can do this, because `sub of a class` is correct. (in this case sub is short for substitute.)

What do you think, @filiphr ?

filiphr replied:
Although "SubclassMapping" might be "more" correct, I don't like it. The biggest reason why I like `SubClassMapping` more is that in an IDE I would always search by Typing "SB" and "SubclassMapping" is not returned in this case. However, reading more about it and looking at some other projects, it seems "Subclass" is more preferred (e.g. in the ByteBuddy project). 

In any case, lets leave the naming for the end. It would be easy to change it. We'll talk about it with the team.

filiphr replied:
Discussed this in the team and we decided to go with `SubclassMapping`. I can do that change.

mjustin commented on documentation/src/main/asciidoc/chapter-10-advanced-mapping-options.asciidoc:
```diff
@@ -110,6 +110,34 @@ public interface SourceTargetMapper {
+----
+====
+
+If you would just use a normal mapping both the AppleDto and the BananaDto would be made into a Fruit object, instead of an Apple and a Banana object. By using the SubClassMapping an AppleDtoToApple mapping will be done for Apples, and an BananaDtoToBanana mapping will be done for Banana's. If you would try to map a GrapeDto it would still turn into a Fruit. In the case that the Fruit type is an  ...
```
Minor, but the plural should be "bananas", not "banana's".

```suggestion
If you would just use a normal mapping both the AppleDto and the BananaDto would be made into a Fruit object, instead of an Apple and a Banana object. By using the SubClassMapping an AppleDtoToApple mapping will be done for Apples, and an BananaDtoToBanana mapping will be done for Bananas. If you would try to map a GrapeDto it would still turn into a Fruit. In the case that the Fruit type is an abstract class or an interface, you would get an `IllegalArgumentException` because it is unknown how to map a GrapeDto. Adding the missing (`@SubClassMapping`) for it will fix that.
```

Also, since the Apple/Banana/etc. types are (hypothetical) Java types, should they be done using code blocks?  Or is it better that they remain the normal code formatting?

Zegveld replied:
Also spotted a few other things just now that can be worded a bit better. (Like "Apples" should be "`AppleDto` objects", as "Banana's" should have been "`BananaDto` objects") 
Will also use the code blocks for the hypothetical types and methods. Think that would make it clear whether something is java code related or not. (I'll add a thumbs up after updating it)

filiphr commented:
>Only done the bits so far where I've added a thumbs up.
Still got a few review items to go, but that will have to wait till when I have time again later this week.

No problem @Zegveld, take as much time as you need. Please ping me once the PR is fully ready for the next iteration of review.

Zegveld commented on processor/src/main/resources/org/mapstruct/ap/internal/model/BeanMappingMethod.ftl:
```diff
@@ -24,6 +24,19 @@
+        else {
+    </#if>
+    <#if isAbstractReturnType()>
+        throw new IllegalArgumentException("Not all subclasses are supported for this mapping. Missing for " + ${parameters[0].name}.getClass());
```
changed it so it uses the same value as the subclassmappings themselves do by getting the parameter name from the first subclassmapping. That way it will be the correct parameter.

On another note, no idea how to find the correct method in the following case:
```java
@SubClassMapping(....)
@Mapping(...., source="otherSource.value")
Target mapSuperClass(SourceSuperClass source, OtherSource otherSource);

@Mapping(....)
Target mapSubClass(SourceSubClass source, OtherSource otherSource);
```
didn't write out the whole thing, but my primary tests show that I'll generate a new method.
`Target mapSourceSubClassToTarget(SourceSubClass source)`
Instead of using the method `mapSubClass` with the additional parameter.

Although my suggestion to those who need this will be create a record containing those 2 arguments so that you only have to pass 1 argument. That way mapstruct will be able to handle it correctly.

filiphr replied:
Multi argument methods are not reused on our side. So you are correct not to focus on that.

Zegveld commented:
@filiphr , Think I've done all the adjustments. So that part is ready for a next iteration of review. Hopefully I haven't missed anything.

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/source/SubClassMappingOptions.java:
```diff
@@ -0,0 +1,116 @@
+        boolean isConsistent = true;
+
+        boolean isChildOfAParameter = false;
+        for ( VariableElement parameter : method.getParameters() ) {
```
Doing `method.getParameters()` is not quite correct since this will also take `@Context` parameters into consideration.  You can pass the real source parameters from the `MethodRetrievalProcessor`, since we have them there. I think you can use them there. There is also `targetParameter` and the `resultType` there as well.

filiphr reviewed (commented):
Great work @Zegveld. We are closing to the end 😄 . I've left some more comments inline. Will reply to yours and resolve what has been already done by you.

Thanks for bearing with us

Zegveld commented on documentation/src/main/asciidoc/chapter-10-advanced-mapping-options.asciidoc:
```diff
@@ -110,6 +110,40 @@ public interface SourceTargetMapper {
+
+[NOTE]
+====
+Combining `@SubClassMapping` with update methods is not supported. If you use an update method with a return value then the subclass mappings will return a new instance based on the mapping, but it will not take into account the `@MappingTarget` parameter.
+The same issue exists for the `@Context` and `@TargetType` parameters.
```
But then you would also need to check if the `@MappingTarget` is of the same type as the `@SubClassMapping(target=TargetSub.class, source=SourceSub.class)` so you would end up with 2 instance of checks in order to call the subclass method which would then become: `void updateTargetSubWithSourceSub(@MappingTarget TargetSub target, SourceSub source)`. So the implementation of the mapping method calling that update method would look something like this:
```java
void updateMapping( @mappingTarget Target target, Source source ) {
    if ( source instanceof SourceSub && target instanceof TargetSub ) {
        updateTargetSubWithSourceSub( (TargetSub) target, (SourceSub) source );
    } else if ( source instanceof SourceSub) {
        // what now??? source argument matches subclass of sourceClass but target argument doesn't match subclass of targetClass.
    } else {
        // default update behavior
    }
}
```
But how to search for an existing method or method definition that can be called?

I think it would indeed be best to have that as another Issue: support for subclassmappings in combination with update mappings. Should I create an issue for this so that we can discuss this topic in that issue's thread?

Zegveld commented on processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:
```diff
@@ -184,7 +196,9 @@ else if ( userDefinedReturnType != null ) {
                     initializeFactoryMethod( returnTypeImpl, selectionParameters );
-                    if ( factoryMethod != null || canReturnTypeBeConstructed( returnTypeImpl ) ) {
+                    if ( factoryMethod != null
+                        || hasSubClassMappingsAndIsEitherAbstractOrCanBeConstructed( returnTypeImpl )
```
That method is still used, inside the others.

Basically there are 2 situations that can happen now.
1. No subclass annotation and return type can be constructed.
2. With subclass annotation and an abstract class or interface as return type or a return type that can be constructed.

So got 2 methods for that now:
1. `doesNotHaveSubClassMappingsAndCanBeConstructed( returnTypeImpl )`
method implementation: `!hasSubclassMappings() && canReturnTypeBeConstructed( returnTypeImpl )`
2. hasSubClassMappingsAndIsEitherAbstractOrCanBeConstructed
method implementation: `hasSubclassMappings() && isReturnTypeAbstractOrCanBeConstructed( returnTypeImpl )`

The methods `isReturnTypeAbstractOrCanBeConstructed` and `canReturnTypeBeConstructed` are the ones that also handle the error message stuff.

Couldn't think of a shorter name that would explain it correctly. So if you have better method suggestions I'm all for it.

Zegveld commented on processor/src/main/java/org/mapstruct/ap/internal/processor/MethodRetrievalProcessor.java:
```diff
@@ -319,6 +342,10 @@ private SourceMethod getMethodRequiringImplementation(ExecutableType methodType,
     }
 
+    private List<SubClassMappingOptions> sortSubClassMappings(Set<SubClassMappingOptions> subClassMappingOptions) {
+        return subClassMappingOptions.stream().sorted( new SubClassInheritenceSorter() ).collect( Collectors.toList() );
```
True, the current implementation doesn't do any sorting atm, but just converts it into a list. 
I noticed that a `LinkedHashSet` is used, I guess that the reason for this is to maintain the user defined order while keeping it distinct?
With a `Set` my assumption would be that the order does not matter. Wouldn't it then be better to explicitly use the `LinkedHashSet` to identify the collection instead of a `Set`?

Zegveld replied:
Made it explicit that the Set is a LinkedHashSet and not a Set with undefined ordering.
Also removed the sorter because it is not needed anymore.

Zegveld commented:
@filiphr made changes to handle 2 of the issues (marked with the thumbs up), but got some questions about the other 3, see the comments per issue.
I will also do the rename from SubClass to Subclass, that is no problem.

Zegveld commented:
Just read the first review again and noticed the bit about:
`Can we also add some fixture tests, to see how the generated code looks like? As you know we strive to generate code which looks human readable smile .`
Where can I find an example of fixture tests so that I can add them for the subclass mappings as well.

Also added a warning for `Something that could be good, but maybe we need to do it in a separate PR is to validate that the SubClassMapping make sense. For example if the first sub class mapping does not hide the second one.`

Zegveld commented:
I've added fixture tests. Made these separate so that the main variants are covered and not impacted by changes within the general tests. 
@filiphr, hope to hear from you soon for the 2 remaining issues, and also what you think of my solution for the annotation handling order one.

sjaakd reviewed (changes requested):
First of all: what a tremendous feat of work. Thanks!. When discussing this with @filiphr I was amazed by the effort done on this.

I did not check the complete implementation. I recon that @filiphr did that already. So I focussed on the API extensions and the implications there.

Now: we do like to go more into meta-annotations in 2.0. So, isolating functionality (`sourceClass` and `targetClass`) into a new annotation is a good idea. I've one consideration here though. 

MapStruct recognises a few major classes of mappings: mapping bean-to-bean, mapping iterable-to-iterable, map-to-map and value-to-value. BeanMapping obviously the most important one. Also, each comes with its own annotation for doing configuration: `@BeanMapping`, `@IterableMapping`, `@MapMapping`, `@ValueMapping`.

There is check functionality that guarantees that  a `MapMapping` is not accidentally applied to `@BeanMapping`. In fact, we do have a small problem with the just released map-to-bean-mapping, which under some conditions gets mixed up with bean mapping.

Having said that: would it not be better to make `@SubclassMapping` a property of `@BeanMapping`? Although more verbose, it would keep functionality together and make sure `@SubclassMapping` is not accidentally applied to another form of mapping.

In 2.0 we could give those mappings a more " independent life", perhaps making even use of the fact that they are part of a `@BeanMapping`. But lets do that step-by-step. 

Next, I was thinking how this functionality would interplay with the new `MapToBean` functionality. Do we have an idea 'bout that?

anyway @Zegveld : what a nice piece of work on an issue which has a long history. Really looking forward to work with you on more issues in the future.

filiphr commented:
Thanks for the review @sjaakd. However, I have to say that I do not quite agree with moving `@SubclassMapping` to the `BeanMapping`. I believe that `@SubclassMapping` is on the same level as `@ValueMapping` and `@Mapping`.  

Adding it to the `@BeanMapping` to only remove it later on doesn't bring much value I think. @Zegveld also did a really nice refactoring which makes supporting meta annotations work out of the box already.

> and also what you think of my solution for the annotation handling order one.

@Zegveld you mean the refactoring you did, right? I found it really great, already said that in https://github.com/mapstruct/mapstruct/pull/2512#discussion_r703782441 😉 .

-----
@Zegveld I've been a bit busy, so I haven't gotten around in finishing the review with your last changes. I haven't forgotten about this one yet.

Zegveld commented:
> First of all: what a tremendous feat of work. Thanks!. When discussing this with @filiphr I was amazed by the effort done on this.
> 
> I did not check the complete implementation. I recon that @filiphr did that already. So I focussed on the API extensions and the implications there.
> 
> Now: we do like to go more into meta-annotations in 2.0. So, isolating functionality (`sourceClass` and `targetClass`) into a new annotation is a good idea. I've one consideration here though.
> 
> MapStruct recognises a few major classes of mappings: mapping bean-to-bean, mapping iterable-to-iterable, map-to-map and value-to-value. BeanMapping obviously the most important one. Also, each comes with its own annotation for doing configuration: `@BeanMapping`, `@IterableMapping`, `@MapMapping`, `@ValueMapping`.
>
> There is check functionality that guarantees that a `MapMapping` is not accidentally applied to `@BeanMapping`. In fact, we do have a small problem with the just released map-to-bean-mapping, which under some conditions gets mixed up with bean mapping.
> 
> Having said that: would it not be better to make `@SubclassMapping` a property of `@BeanMapping`? Although more verbose, it would keep functionality together and make sure `@SubclassMapping` is not accidentally applied to another form of mapping.
> 

I didn't think about this initially, but it would be possible to do submappings for iterables or maps as well. This would be useful if the mapping is reused multiple times, but the implementation collection varies. And you could even support implementations of the Collection framework that you do not know exist. (e.g. project specific ones) 
For `@ValueMapping`s I'm not that certain, if this is only enums then no. But if this is also static fields than the same thing could apply. Because static fields can refer to subclassed instances of the source. 
It would require some more work to support the `@IterableMapping`, `@MapMapping` and `@ValueMapping` annotations with subclasses but I do believe that would be possible. Not certain as to how often it would be used if at all though, because the above is pure theoretical.

> In 2.0 we could give those mappings a more " independent life", perhaps making even use of the fact that they are part of a `@BeanMapping`. But lets do that step-by-step.
> 
> Next, I was thinking how this functionality would interplay with the new `MapToBean` functionality. Do we have an idea 'bout that?

I think that could turn out to be an adjusted variant of the `@SubclassMapping` as is, but instead of supplying a source class you would supply a sourceCondition. If this sourceCondition would result in a true value a more specific MapToBean method would be called. Something for in another PR but here is the gist of it:
```
@SubclassMapping( sourceCondition="java(input.containsKey('subBeanSpecificField'))", target=SubBean.class )
Bean mapToBean(Map<String, String> input);

@Mapping(/*specific mapping for this condition*/
SubBean mapToSubBean(Map<String, String> input);
```
which could result in something like:
```
Bean mapToBean(Map<String, String> input) {
    if (input.containsKey('subBeanSpecificField')) {
        return mapToSubBean(input);
    }
    // default mapToBean behaviour
}

SubBean mapToSubBean(Map<String, String> input) {
    // mapToSubBean behaviour
}
```
> anyway @Zegveld : what a nice piece of work on an issue which has a long history. Really looking forward to work with you on more issues in the future.

Likewise.



> > and also what you think of my solution for the annotation handling order one.
>
> @Zegveld you mean the refactoring you did, right? I found it really great, already said that in #2512 (comment) wink .

@filiphr, no I did a refactor which changes several Set's into specifically LinkedHashSet's. Was talking about that one.

> @Zegveld I've been a bit busy, so I haven't gotten around in finishing the review with your last changes. I haven't forgotten about this one yet.

No problem I've got plenty of waiting time. Currently implementing just a lot of instanceof checks manually when I'm using it instead of generating them. :)

sjaakd commented:
> I didn't think about this initially, but it would be possible to do submappings for iterables or maps as well. This would be useful if the mapping is reused multiple times, but the implementation collection varies. And you could even support implementations of the Collection framework that you do not know exist. (e.g. project specific ones)

You have a point here. Did not consider that. I think @filiphr and I were discussing the change of the default implementation (don't remember if it was `Map` or `List`) a while ago.


> For `@ValueMapping`s I'm not that certain, if this is only enums then no. But if this is also static fields than the same thing could apply. Because static fields can refer to subclassed instances of the source.

That would be really a niche feature. 

I was discussing mis-use of annotations with @filiphr just earlier and he pointed out that we don't have (presumably) checks for `@Mapping` either. So we're a bit half-hearted with checks here anyway. 

Finally, applying it to the new map-to-bean would indeed qualify as a follow up PR.

So, as far as I'm concerned: go ahead.

filiphr commented:
I checked out the PR locally today and did some changes in it. I've pushed the changes to https://github.com/filiphr/mapstruct/tree/feature/PR-131 (I couldn't push it to your branch for some reason, most likely you have not allowed maintainers to push changes to it). Can you please have a look at them and let me know whether you agree with them @Zegveld? I tried to do them atomically in order to show my thought process for some of them, everything will be merged into one commit in the end.

>@filiphr, no I did a refactor which changes several Set's into specifically LinkedHashSet's. Was talking about that one.

I saw that one, and I didn't really like it. I do not like having collection implementation types as fields of classes or parameters. I reverted some of that change in a separate PR.

Some other more notable changes I did are:

* We do not support all update methods, not just the ones that are void. Otherwise it is too confusing
* Small Refactoring in the RepeatableMappingAnnotations to reduce the complexity a bit
* Remove obsolete changes in MapperCreationProcessor

Some final things to address before merging:

* I am not sure that the throwing of the `IllegalArgumentException` is a good approach. The reason for that is that if you remove the `SubclassMapping`(s) we will have a compile error, otherwise we throw a runtime error. Can we keep this behaviour consistent? In a follow up PR we can add support for throwing the exception if the user wants it by explicitly asking for it via `@BeanMapping` and in another PR to support sealed classes, i.e. we can know whether the `SubclassMapping`(s) have all classes for the source or not.

I think that this would be the last change I request, together with my changes the PR would be ready for merging and doing a beta release shortly after that. It would be a great addition to the 1.5 release 😄 

------

Side note: Since your email address is private, can you please reach out to me somehow? You could use [Twitter](https://twitter.com/filiphr) or my email address (it is in my commits). I'd like to discuss something with you.

Zegveld commented:
I did take that route under consideration, but the use case that I had which caused me to write this enhancement was one where I never have an implementation for general usage. In this use case the classes are generated from two xsd's and I need to transfer information between the two models.

The best solution I can think of usage side without having the exception would be to manually create an implementation/extension of the base class which then would throw an IllegalStateException in it's constructor because it should not be created. That one would then be the default implementation that mapstruct falls back to. This does not feel that good as a solution for this problem.

What if we add another annotation `@SubclassMappingStrategy` in which you can configure whether you get an exception or compile error. Hmmm maybe with a field/property named .... can't think of a good name for the property atm. But that property would then influence the behavior. Just like how the other MappingStrategy annotations do.

Zegveld commented:
@filiphr , I've introduced the SubclassExhaustiveStrategy and did some more renaming/refactoring. Also updated the documentation to have it mentioned.
Hope you have time to review the new changes.
